### PR TITLE
Triangle subdivision

### PR DESCRIPTION
**Pull request by stla**
Is it normal that we see the hidden edges with CairoMakie?

ptqrefine-Float64

### CODE REVIEW

**Comment by juliohm** [comment]
CairoMakie can only fake 3D. I personally never understood the rationale
behind this backend. It kind of goes against the whole Makie design as an
interactive 3D visualization system.

Em sáb., 24 de dez. de 2022 22:59, stla ***@***.***> escreveu:

File: src/Meshes.jl
Comment (juliohm) [inline src/Meshes.jl]: We prefer more descriptive and explicit names in Julia. Is there a standard name for this method? Ideally we retain the name of the original paper that introduced it.

File: src/refinement/ptq.jl
Comment (juliohm) [inline src/refinement/ptq.jl]: Please try to follow the docstring convention of other methods. In particular, it is very helpful when we provide references in the literature. Is there a published paper discussing/using this method?

File: src/refinement/ptq.jl
Comment (juliohm) [inline src/refinement/ptq.jl]: Are there plans to support other types of "middle" function?  I think this function primal_triangle_quadrisection could be the refine function. No need to introduce a separate function.

**Review by juliohm — CHANGES_REQUESTED** [review]
Review attached.

File: src/refinement/ptq.jl
Comment (juliohm) [inline src/refinement/ptq.jl]: Can we generalize this refinement method to support arbitrary polygonal meshes?  I think we can call it EarRefinement, an algorihtm that clips ears of the polygon as triangles and then triangulates the interior polygon with FanTriangulation.

File: src/Meshes.jl
Comment (stla) [reply src/Meshes.jl]: It is called PTQ in several papers, and in CGAL.

File: src/refinement/ptq.jl
Comment (stla) [reply src/refinement/ptq.jl]: Yes, the hyperbolic midpoint to construct hyperbolic triangles. Also if you take three points on the unit sphere and you iterate the PTQ with the normalized midpoint, you get a spherical triangle.

File: src/refinement/ptq.jl
Comment (stla) [reply src/refinement/ptq.jl]: By calling `triangulate()` at the beginning?

File: src/Meshes.jl
Comment (stla) [reply src/Meshes.jl]: And this a subdivision algorithm of the so-called family of **loop** subdivision algorithms.

File: src/refinement/ptq.jl
Comment (stla) [reply src/refinement/ptq.jl]: The CGAL documentation refers to a master thesis:

Charles Loop. *Smooth subdivision surfaces based on triangles. Master's thesis, University of Utah, 1987.*

Lol, I thought that the world "loop" was refering to a for-loop...

File: src/Meshes.jl
Comment (juliohm) [reply src/Meshes.jl]: > It is called PTQ in several papers, and in CGAL.

The R language calls many things, ABC, DET, PTYS, ..., but this is not necessarily a good practice. It makes things unsearchable and the Julia community is working really hard to avoid unnecessary acronyms everywhere. We should rename it to a more descriptive searchable name. Charles Loop's thesis describes a couple of algorithms in chapter 3, which one is the one in this PR? https://charlesloop.com/thesis.pdf

File: src/refinement/ptq.jl
Comment (juliohm) [reply src/refinement/ptq.jl]: Awesome, we should definitely add this his thesis to the docstring.

File: src/refinement/ptq.jl
Comment (juliohm) [reply src/refinement/ptq.jl]: We can refactor our refinement methods later to add this flexibility and avoid code duplication. As far as I remember our CatmullClark method is the same as TriRefinement with a custom "middle" function as well.

File: src/refinement/ptq.jl
Comment (juliohm) [reply src/refinement/ptq.jl]: That would be an interesting alternative as well.

File: src/Meshes.jl
Comment (stla) [reply src/Meshes.jl]: I never talked about R here. I didn't see any name in the thesis. I first found the name *Primal Triangle Quadrisection* in the CGAL documentation, and googling this name yields some results.

Comment (stla) [reply src/Meshes.jl]: Maybe `TriQuad`? For "triangle quadrisection"?

Comment (juliohm) [reply src/Meshes.jl]: > I never talked about R here.

I know :) Just pointing out that acronyms are terrible choices and we should avoid whenever possible. We use them sometimes, but it is worth checking if there is a more descriptive name to (1) honor the author of the algorithm and (2) make the code more searchable.

> I didn't see any name in the thesis.

image

His chapter 3 describes a couple of algorithms. You can see that he mentions Catmull-Clark there, which we already have implemented here. Maybe the algorithm you implemented is one of the other ones?

Comment (juliohm) [reply src/Meshes.jl]: > Maybe `TriQuad`? For "triangle quadrisection"?

I think it is a bit confusing. I read it as a refinement procedure that produces triangles and quadrangles, but that is not the case here.

Comment (stla) [reply src/Meshes.jl]: `TriQuadrisectionRefinement`?

File: src/refinement/triquadrisection.jl
Comment (juliohm) [inline src/refinement/triquadrisection.jl]: We shouldn't overwrite user input. I think we need to save the result in a new variable, maybe tmesh

File: src/refinement/triquadrisection.jl
Comment (juliohm) [inline src/refinement/triquadrisection.jl]:
```suggestion

  # convert to half-edge structure
  t = convert(HalfEdgeTopology, topology(mesh))
```

File: src/refinement/triquadrisection.jl
Comment (juliohm) [inline src/refinement/triquadrisection.jl]: rename edge_index to eind, or edge, or ind, ... some name that doesn't have underscores

File: src/refinement/triquadrisection.jl
Comment (juliohm) [inline src/refinement/triquadrisection.jl]:
```suggestion
_ordered(i ,j) = i < j ? (i, j) : (j, i)
```

File: src/refinement/triquadrisection.jl
Comment (juliohm) [inline src/refinement/triquadrisection.jl]: ❤️

File: test/refinement.jl
Comment (juliohm) [inline test/refinement.jl]:
```suggestion
      @test_reference "data/triquadrefine-$T.png" fig
```

File: src/refinement/triquadrisection.jl
Comment (juliohm) [inline src/refinement/triquadrisection.jl]:
```suggestion
  tmesh = eltype(mesh) <: Triangle ? mesh : simplexify(mesh)
```

File: src/refinement/triquadrisection.jl
Comment (juliohm) [inline src/refinement/triquadrisection.jl]:
```suggestion
    midpoint = center(Segment(points[i], points[j]))
```

File: src/refinement/triquadrisection.jl
Comment (juliohm) [inline src/refinement/triquadrisection.jl]:
```suggestion
  midpoints = Dict{Tuple{Int,Int},Int}()
```

File: src/refinement/triquadrisection.jl
Comment (juliohm) [inline src/refinement/triquadrisection.jl]:
```suggestion
    i, j = sort(∂₁₀(eind))
```

File: src/refinement/triquadrisection.jl
Comment (juliohm) [inline src/refinement/triquadrisection.jl]:
```suggestion
    midpoints[(i,j)] = npoints
```

File: src/refinement/triquadrisection.jl
Comment (juliohm) [inline src/refinement/triquadrisection.jl]:
```suggestion
    m12 = midpoints[_ordered_pair(v1, v2)]
    m23 = midpoints[_ordered_pair(v2, v3)]
    m31 = midpoints[_ordered_pair(v3, v1)]
```

**Review by juliohm — CHANGES_REQUESTED** [review]
We also need to delete old image files pqtrefine-*.png that were generate in a previous version of the code.

File: src/refinement/triquadrisection.jl
Comment (stla) [reply src/refinement/triquadrisection.jl]: `center` is not implemented for `Segment`. Can I implement it in this PR? Or I open a new one? Hmm, yes, I think a new one is better.

File: src/refinement/triquadrisection.jl
Comment (juliohm) [inline src/refinement/triquadrisection.jl]: I think we can avoid the construction of this dictionary here. We can simply rely on the insertion order. Instead of looping over facets, try to loop over elements (i.e. triangles) and for each of these elements, you insert 4 new points to the end of the vector of points. In a second loop you can then lookup the points in the vector with the correct offsets, and that should be even faster and more memory efficient than then dictionary solution.

Comment (stla) [reply src/refinement/triquadrisection.jl]: I'm not convinced. Some faces share a common edge. If I follow the algorithm you outline, if I correctly understand, I will compute multiple times the midpoints of the edges. No?

**Comment by juliohm** [comment]
I've read Loop's thesis and I think we should honor his work by calling this method `LoopSubdivision`

File: src/refinement/triquadrisection.jl
Comment (juliohm) [reply src/refinement/triquadrisection.jl]: You're right. I had the other refinement method in mind where new points lie inside the elements.

File: src/refinement/loopsubdivision.jl
Comment (juliohm) [inline src/refinement/loopsubdivision.jl]: A final question: in the Loop's thesis, the subdivision procedure is similar to Catmull-Clark in the sense that the initial mesh is considered a de Boor net and the iteration process smooths it out with midpoints that are not necessarily on the edges. Perhaps we should do the same here and "smooth" the points.  Alternatively, we could find a different name for the method that is not LoopSubdivision. Which method do you prefer?

At some point in the future we need to refactor these refinement methods to give more flexibility in the way the new vertices are computed. We have a lot of code repetition currently. For example, QuadRefinement and CatmullClark are basically the same algorithm, except for the new vertex expressions.

**Comment by stla** [comment]
I took a look at the Loop subdivision, e.g. [Loop.cpp](https://github.com/icemiliang/loop_subdivision/blob/master/src/LOOP.cpp). Does it look doable in Meshes.jl?

**Comment by juliohm** [comment]
> I took a look at the Loop subdivision, e.g. [Loop.cpp](https://github.com/icemiliang/loop_subdivision/blob/master/src/LOOP.cpp). Does it look doable in Meshes.jl?

I think we have more generic interfaces for some of the implementation details. Your PR here is almost ready, I think we just need to update the coordinates of the vertices like Loop did. The subdivision itself is already implemented as far as I remember.

**Comment by stla** [comment]
The subdivision is far to be enough I think. I simply took the middle of the edges. But the construction of the locations of the vertices of the Loop subdivision is much more complicated. For each edge it depends on the two faces adjacent to this edge (if this is not a border edge).

**Comment by juliohm** [comment]
> The subdivision is far to be enough I think. I simply took the middle of the edges. But the construction of the locations of the vertices of the Loop subdivision is much more complicated. For each edge it depends on the two faces adjacent to this edge (if this is not a border edge).

Yes, we have a similar vertex calculation in our CatmullClark subdivision algorithm. Maybe you could get inspiration from there? The main challenge is understand how Loop computes the vertex, the implementation should be easy with our topological relations.

**Comment by juliohm** [comment]
Alternatively, we could rename this algorithm here to something else that is not Loop's subdivision. Our current list of refinement methods is:

1. TriRefinement: converts n-gons into triangles
2. QuadRefinement: converts n-gons into quadrangles
3. CatmullClark: same as above + smoothing

Given that the current implementation in this PR always splits n-gons into triangles we need a name that distinguishes it from the existing TriRefinement.

File: src/refinement/loopsubdivision.jl
Comment (juliohm) [inline src/refinement/loopsubdivision.jl]:
```suggestion
function refine(mesh, ::LoopSubdivision)
  @assert paramdim(mesh) == 2 "LoopSubdivision only defined for surface meshes"
  
```

File: src/refinement/loopsubdivision.jl
Comment (juliohm) [inline src/refinement/loopsubdivision.jl]:
```suggestion
  TriSubdivision()
```

File: src/refinement/loopsubdivision.jl
Comment (juliohm) [inline src/refinement/loopsubdivision.jl]:
```suggestion
struct TriSubdivision <: RefinementMethod end

function refine(mesh, ::TriSubdivision)
  @assert paramdim(mesh) == 2 "TriSubdivision only defined for surface meshes"
```

File: src/Meshes.jl
Comment (juliohm) [inline src/Meshes.jl]:
```suggestion
  TriSubdivision,
```

File: src/refinement.jl
Comment (juliohm) [inline src/refinement.jl]:
```suggestion
include("refinement/trisubdivision.jl")
```

File: test/refinement.jl
Comment (juliohm) [inline test/refinement.jl]:
```suggestion
  @testset "TriSubdivision" begin
```

File: test/refinement.jl
Comment (juliohm) [inline test/refinement.jl]:
```suggestion
    ref1   = refine(mesh, TriSubdivision())
    ref2   = refine(ref1, TriSubdivision())
    ref3   = refine(ref2, TriSubdivision())
```

File: test/refinement.jl
Comment (juliohm) [inline test/refinement.jl]:
```suggestion
      @test_reference "data/trisubdivision-$T.png" fig
```

**Review by juliohm — CHANGES_REQUESTED** [review]
@stla let's just rename LoopSubdivision to TriSubdivision and rename the corresponding files. After that we can merge right away.

This method is already useful to have. We can come back to LoopSubdivision in the future.

**Review by juliohm — APPROVED** [review]

**Comment by stla** [comment]
The TriRefinement is also useful for the clipping, to add some vertices when there are not enough.

**Pull request merged**
